### PR TITLE
Add chunk eval op

### LINKED ISSUE #4749: Add Operator used as Chunk Evaluator

### PR DESCRIPTION
**Pull request by guoshengCS**
resolves #4749

### CODE REVIEW

File: paddle/operators/chunk_eval_op.cc [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        .SetDefault("IOB");
+    AddAttr<std::vector<int>>(
+        "excluded_chunk_types",
+        "(list<int>) A list<int> indicating chunk types not to be counted.")
+        .SetDefault(std::vector<int>{});
+    AddComment(R"DOC(
+Chunk evaluator is used to evaluate segment labelling accuracy for a
+sequence. It calculates precision, recall and F1 scores for the chunk detection.
+To use chunk evaluator, several concepts need to be clarified firstly.
+[Chunk type] is the type of the whole chunk and a chunk consists of one or several words.  (For example in NER, ORG for organization name, PER for person name etc.)
```
Comment (qingqing01) [inline paddle/operators/chunk_eval_op.cc]: Add an empty line before line 81 and 82.

Give the full name for the `NER`.

File: paddle/operators/chunk_eval_op.cc [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+- Use one of the listed labelling schemes. These schemes differ in ways indicating chunk boundry.
+
+    Scheme    Description
+    plain    Use the same label for the whole chunk.
+    IOB      Two labels for chunk type X, B-X for chunk begining and I-X for chunk inside.
+    IOE      Two labels for chunk type X, E-X for chunk ending and I-X for chunk inside.
+    IOBES    Four labels for chunk type X, B-X for chunk begining, I-X for chunk inside, E-X for chunk end and S-X for single word chunk.
+
+To make it clear, let's illustrate by an NER example.
+Assuming that there are three named entity types including ORG, PER and LOC which are called 'chunk type' here,
```
Comment (qingqing01) [inline paddle/operators/chunk_eval_op.cc]: Explain the LOC here?

File: paddle/operators/chunk_eval_op.cc [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+if 'IOB' scheme were used, the label set will be extended to a set including B-ORG, I-ORG, B-PER, I-PER, B-LOC, I-LOC and O,
+in which B-ORG for begining of ORG and I-ORG for inside of ORG.
+Prefixes which are called 'tag type' here are added to chunk types and there are two tag types including B and I.
+Of course, the training data should be labeled accordingly.
+- Mapping is done correctly by the listed equations and assigning protocol.
+The following table are equations to extract tag type and chunk type from a label.
+
+    tagType = label % numTagType
+    chunkType = label / numTagType
+    otherChunkType = numChunkTypes
```
Comment (qingqing01) [inline paddle/operators/chunk_eval_op.cc]: The `numTagType` and `numChunkTypes` here is clear, but better to explain them again.

File: paddle/operators/chunk_eval_op.h [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    for (int i = 0; i < num_sequences; ++i) {
+      int seq_length = lod[0][i + 1] - lod[0][i];
+      EvalOneSeq(inference_data + lod[0][i], label_data + lod[0][i], seq_length,
+                 output_segments, label_segments, num_output_segments,
+                 num_label_segments, num_correct, num_chunk_types,
+                 num_tag_types, other_chunk_type, tag_begin, tag_inside,
+                 tag_end, tag_single, excluded_chunk_types);
+    }
+    *precision_data =
+        !num_output_segments ? 0 : (T)num_correct / num_output_segments;
```
Comment (qingqing01) [inline paddle/operators/chunk_eval_op.h]: `(T) num_correct`  -> `static_cast<T>`

File: paddle/operators/chunk_eval_op.cc [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    I-PER  3
+    B-LOC  4
+    I-LOC  5
+    O      6
+
+In this example, chunkType has three values: 0 for ORG, 1 for PER, 2 for LOC, because the scheme is
+"IOB" so tagType has two values: 0 for B and 1 for I.
+Here we will use I-LOC to explain the above mapping rules in detail.
+For I-LOC, the label id is 5, so we can get tagType=1 and chunkType=2, which means I-LOC is a part of NER chunk LOC
+and the tag is I.
```
Comment (wanghaoshuang) [inline paddle/operators/chunk_eval_op.cc]: How about giving an example here?

Steven        B-PER     2
Paul            I-PER       3
Jobs           I-PER       3
works         O              6
for               O              6
Baidu          B-ORG     0
Inc.             I-ORG      1
at                O              6
Beijing        B-LOC      4
of                I-LOC       5
China          I-LOC       5

File: paddle/operators/chunk_eval_op.h [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    *racall_data =
+        !num_label_segments ? 0 : (T)num_correct / num_label_segments;
+    *f1_data = !num_correct ? 0 : 2 * (*precision_data) * (*racall_data) /
+                                      ((*precision_data) + (*racall_data));
+  }
+
+  void EvalOneSeq(const int* output, const int* label, int length,
+                  std::vector<Segment>& output_segments,
+                  std::vector<Segment>& label_segments,
```
Comment (wanghaoshuang) [inline paddle/operators/chunk_eval_op.h]: `output_segments` and `label_segments` are not used outside of `EvalOneSeq`. So why not difine them in  `EvalOneSeq` and remove them from arguments list?

Comment (wanghaoshuang) [inline paddle/operators/chunk_eval_op.h]: `output_segments` and `label_segments` are not used outside of `EvalOneSeq`. So why not difine them in  `EvalOneSeq` and remove them from arguments list?

File: paddle/operators/chunk_eval_op.cc [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                         "(string, default IOB) The label scheme.")
+        .SetDefault("IOB");
+    AddAttr<std::vector<int>>(
+        "excluded_chunk_types",
+        "(list<int>) A list<int> indicating chunk types not to be counted.")
+        .SetDefault(std::vector<int>{});
+    AddComment(R"DOC(
+Chunk evaluator is used to evaluate segment labelling accuracy for a
+sequence. It calculates precision, recall and F1 scores for the chunk detection.
+To use chunk evaluator, several concepts need to be clarified firstly.
```
Comment (lcy-seso) [inline paddle/operators/chunk_eval_op.cc]: we first introduce some related concepts.

File: paddle/operators/chunk_eval_op.cc [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+              "(float) The recall ratio of the predictions on current data.");
+    AddOutput("F1-Score",
+              "(float) The F1-Score of the predictions on current data.");
+    AddAttr<int>("num_chunk_types", "(int) The number of chunk type.");
+    AddAttr<std::string>("chunk_scheme",
+                         "(string, default IOB) The label scheme.")
+        .SetDefault("IOB");
+    AddAttr<std::vector<int>>(
+        "excluded_chunk_types",
+        "(list<int>) A list<int> indicating chunk types not to be counted.")
```
Comment (lcy-seso) [inline paddle/operators/chunk_eval_op.cc]: - indicating chunk types that are not counted.
- This explanation is hard to understand for users.

File: paddle/operators/chunk_eval_op.cc [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+    PADDLE_ENFORCE(inference_dim == label_dim,
+                   "Inference's shape must be the same as Label's shape.");
+
+    ctx->SetOutputDim("Precision", {1});
+    ctx->SetOutputDim("Recall", {1});
+    ctx->SetOutputDim("F1-Score", {1});
+  }
+
+  framework::DataType IndicateDataType(
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Done.

File: paddle/operators/chunk_eval_op.cc [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  }
+};
+
+class ChunkEvalOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  ChunkEvalOpMaker(framework::OpProto *proto,
+                   framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Inference",
+             "(Tensor, default: Tensor<int>) Predictions from the network.");
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Done.

File: paddle/operators/chunk_eval_op.cc [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    AddInput("Label", "(Tensor, default: Tensor<int>) Labels of the data.");
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Done.

File: paddle/operators/chunk_eval_op.cc [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    AddOutput(
+        "Precision",
+        "(float) The precision ratio of the predictions on current data.");
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Done.

File: paddle/operators/chunk_eval_op.cc [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    AddOutput("Recall",
+              "(float) The recall ratio of the predictions on current data.");
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Done.

File: paddle/operators/chunk_eval_op.cc [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    AddOutput("F1-Score",
+              "(float) The F1-Score of the predictions on current data.");
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Done.

File: paddle/operators/chunk_eval_op.cc [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    AddAttr<int>("num_chunk_types", "(int) The number of chunk type.");
+    AddAttr<std::string>("chunk_scheme",
+                         "(string, default IOB) The label scheme.")
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Done.

File: paddle/operators/chunk_eval_op.cc (hunk 8 above)
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Done. Add `see below for details`.

File: paddle/operators/chunk_eval_op.cc [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        .SetDefault("IOB");
+    AddAttr<std::vector<int>>(
+        "excluded_chunk_types",
+        "(list<int>) A list<int> indicating chunk types not to be counted.")
+        .SetDefault(std::vector<int>{});
+    AddComment(R"DOC(
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

File: paddle/operators/chunk_eval_op.cc [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+Chunk evaluator is used to evaluate segment labelling accuracy for a
+sequence. It calculates precision, recall and F1 scores for the chunk detection.
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

File: paddle/operators/chunk_eval_op.cc (hunk 7 above)
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

File: paddle/operators/chunk_eval_op.cc (hunk 1 above)
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

File: paddle/operators/chunk_eval_op.cc [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+To use chunk evaluator, several concepts need to be clarified firstly.
+[Chunk type] is the type of the whole chunk and a chunk consists of one or several words.  (For example in NER, ORG for organization name, PER for person name etc.)
+[Tag type] indicates the position of a word in a chunk. (B for begin, I for inside, E for end, S for single)
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

File: paddle/operators/chunk_eval_op.cc (hunk 2 above)
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

File: paddle/operators/chunk_eval_op.cc (hunk 3 above)
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

File: paddle/operators/chunk_eval_op.cc (hunk 5 above)
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.cc]: Rewrite the doc.

File: paddle/operators/chunk_eval_op.h (hunk 4 above)
Comment (guoshengCS) [reply paddle/operators/chunk_eval_op.h]: Done.

**Pull request merged**